src/assets/TokenBalancesController.test.ts

```diff
@@ -115,115 +121,136 @@ describe('TokenBalancesController', () => {
     });
   });
 
-  it('should update all balances', async () => {
-    const network = new NetworkController();
-    const preferences = new PreferencesController();
-    const assets = new TokensController({
-      onPreferencesStateChange: (listener) => preferences.subscribe(listener),
-      onNetworkStateChange: (listener) => network.subscribe(listener),
+  describe('with network controller', () => {
+    let messenger: NetworkControllerMessenger;
+    let preferences: PreferencesController;
+    beforeEach(() => {
```

Comment:
using beforeEach here... Not sure what the cutoff is / should be for when we use beforeEach vs. not. To me, anytime we want to do cleanup, we should use beforeEach/afterEach so that down the road, devs don't need to remember/know about what they should clean up in their new tests.

Reply:
@BelfordZ My feeling is that we should avoid this pattern as much as possible, because of the reasons I mentioned previously. Even if the setup itself looks innocuous now, it doesn't take long before you have to rewrite this into a more maintainable form. I think it would be nice if we could follow patterns when writing tests that 1) work both for simpler tests and more complicated tests and 2) don't force us to spend precious time refactoring tests when they reach a certain level of complexity. So my vote would be to extract this setup into some kind of function.

Reply:
>To me, anytime we want to do cleanup

This is similar to how I think of it as well. I use `afterEach` whenever there is some cleanup to do. Not just to ensure devs don't forget the cleanup step, but also so that we can guarantee that the cleanup runs if the test fails (which otherwise would require an awkward `try { ... } finally { ... }` wrapper around every test with cleanup, awful for readability).

And then if the cleanup in `afterEach` is referencing a variable (like a controller) rather than a global, it makes sense to setup that variable in `beforeEach` so that the reader can understand what the `afterEach` step is doing, and so that we don't accidentally forget to set the variable being cleaned up.

That pattern does come at a readability cost, in that the setup has moved from the test to the setup. This can make it more difficult to read, as you'd need to scroll back and forth between the test and setup to keep the context in your head. But it can be a reasonable tradeoff in my opinion.

There is an alternative pattern that @mcmire favors where we do cleanup in a wrapper function instead. You can create a setup function (like `withController`) that takes a callback, which gets passed the things that need cleanup. Then the setup function has a finally block that cleans things up correctly without complicating the test. This also comes at a readability cost, in that if the setup function is complicated enough that you need to dive in to understand it, you're left with the same scrolling-back-and-forth problem. It also requires an additional level of indentation, because the test itself is within the callback. But it can also be a reasonable tradeoff, especially with a setup function that is really easy to understand the purpose and behavior of. @mcmire 's recent extension tests using the `withInfuraClient` are a good example of how this pattern can abstract away a lot of complex setup in a way that still makes it easy to understand what is happening in the test.

I might at some point be convinced that the wrapper function pattern is better in all cases, but today I think both can be pretty decent solutions, depending on the specific case.

Reply:
The more clear-cut case where `beforeEach` has a strictly better alternative is when we're _just_ doing setup, and there is no cleanup to do. In my opinion it's strictly better for readability if test setup is done directly in the test or via a setup function called at the beginning of the test, versus being in `beforeEach`. It's the same result with less scrolling.

Reply:
The most egregious maintenance problems we've seen from using `beforeEach` are cases where the setup in `beforeEach` no longer matches what each tests needs. For example, it could have some steps that are needed by a subset of tests, and they're totally unused by some tests. Or it could have setup steps that are undone within the test itself. This is awful and confusing.

This isn't an inherent problem with `beforeEach`, but it's easy to see how people slip into this anti-pattern when we default to using `beforeEach` for all setup. So it's another reason to be wary of using it for setup.

Reply:
Ah, I see this `with__` pattern was [covered already in a separate comment](https://github.com/MetaMask/controllers/pull/903#discussion_r980519671), and that there is a tracking issue for this now (https://github.com/MetaMask/controllers/issues/922)

Reply:
> The more clear-cut case where beforeEach has a strictly better alternative is when we're just doing setup

That's a good point. I do tend to favor beforeEach/afterEach for things I know are short and sweet, like:

* `jest.useFakeTimers()` / `jest.useRealTimers()`
* `nock.cleanAll()`

It doesn't really feel like we are adding complexity by doing this.